### ISSUE TITLE
Nested Values Disappear when using StrictMode and `pushValue`

### ISSUE THREAD

**Issue by dlindahl** [TanStack/form#704]
### Describe the bug

> @dlindahl I believe what you're describing is a separate issue that pertains to pushValue. Can you make a new issue?

_Originally posted by @crutchcorn in https://github.com/TanStack/form/issues/685#issuecomment-2089490042_
            
---

I am seeing an issue with React's `StrictMode` and `pushValue` with an array field where it doesn't appear to do anything. Under the hood, I believe it is adding the value as expected but then immediately reverting the field value to whatever the previous value was.

### Your minimal, reproducible example

https://stackblitz.com/edit/tanstack-form-rxy9kq?file=src%2Findex.tsx&preset=node

### Steps to reproduce

I took the array example from the docs and changed it in the following ways:

1. Wrap App in `StrictMode`.
2. Change `[Array<{ name: string; age: number }>](people: [] as Array<{ name: string; age: number }>)` to `people: [] as Array<String>`
3. Clicking on the "Add person" button does not result in a new row.

### Expected behavior

As a user, I expect the "Add person" button to push a new value to the field's internal array state.

### How often does this bug happen?

Every time

### Platform

Chrome 123.0.6312.87 (Official Build) (arm64)

### TanStack Form adapter

None

### TanStack Form version

v0.19.4

### Additional context

Also of note is when you leave the people value to be an array of objects, `StrictMode` works fine. Its only when the underlying value is an array of strings that it breaks. If I had to guess, I'd say there is a value check somewhere that is failing since an empty string is "falsey", but I have no idea what the root cause would be.

**Comment by Balastrong** [comment]
I'm not sure _why_ this is happening, but looks like the `Field` components get unmounted and in the callback here we delete the field.

https://github.com/TanStack/form/blob/3e827afac920ba312691670a93724edc9fcfed2b/packages/form-core/src/FieldApi.ts#L403-L409

In fact, if you add `preserveValue` in `<form.Field key={i} name={`people[${i}]`} preserveValue>` this behavior is skipped and the values are getting pushed as expected.

This seems to be the same root cause as #699

**Comment by crutchcorn** [comment]
@Balastrong I think you're on the right track here, but the problem is that `deleteField` and creating a field should create an equilibrium in effects if ran back-to-back, ala StrictMode.

If you're able to get to this before I'm able to investigate, let's see if we can iron out _why_ creating a field with numbers and nested values disappears. Look at the mounting behavior, not the unmounting behavior.

**Comment by crutchcorn** [comment]
The problem that's occuring here is incredibly specific to both:

- Our unique framework agnostic tech stack
- The way StrictMode works with dual-rendering
- The way the VDOM works with React

TLDR what's happening is that:

- We are `mount`ing the `people[0]` `FieldAPI`, which adds it to the `FormAPI.state.values` array
- We then call the `mount` `unsubscribe` on the `people[0]` `FieldAPI`, which removes it from `FormAPI.state.values` array
- This in turn notifies the `people` `FieldAPI`, which triggers the change on the array itself and re-renders the VDOM list
- This VDOM list is updated and therefore _unmounts_ `people[0]`, which runs the cleanup on the `FieldAPI` on `people[0]`, which is why it never renders

To solve this, I think we need to introduce a notification priority system with a bottom-up notification tree (notify `people[0]`, then `people`, then `''`).

**Comment by crutchcorn** [comment]
Disregard, the challenge is that `defaultValue` is undefined, which never triggers this line of code:

https://github.com/TanStack/form/blob/main/packages/form-core/src/FieldApi.ts#L448-L450

As a result, the code goes through the following processes:

1a) First (dummy StrictMode) render:
    i) Sets up the `FieldAPI` for `people` and:
    ii) `people[0]`
1b) First (dummy StrictMode) cleanup:
    iii) Deletes `people[0]`
    iiii) Deletes `people`, removing the array from `FormAPI.store.state.values`
2a) Second (real) render:
    v) Creates `people[0]` _first_, but does not trigger the `inform the form` creation
    vi) _Then_ handles `people`, which recreates the (now empty) array 

Because the `people` array is now empty, it creates a:

3a) Second (real) cleanup

To solve this temporarily, simply add:

```
defaultValue={""} 
```

To here:

```tsx
<form.Field key={i} defaultValue={""} name={`people[${i}]`}/>
```

PR with fix soon

**Comment by chidiwilliams** [comment]
I've noticed the fix still doesn't work for nested objects within the array in StrictMode.

Taking the original example from the docs and wrapping with Strict Mode, adding a new Person object three times leads to: `[{"age":0},{"age":0},{"name":"","age":0}]`.

https://stackblitz.com/edit/tanstack-form-ien81g?file=src%2Findex.tsx

**Comment by crutchcorn** [comment]
@dlindahl that warning comes from TanStack React Form and IDK why. It can be safely ignored tho

**Comment by crutchcorn** [comment]
@chidiwilliams let's make a new issue. I think this is a different underlying problem, even tho it looks very similar

**Comment by dlindahl** [comment]
@crutchcorn I'll do some more debugging on my end but I'm not _100%_ sure it can be ignored. In my very specific implementation of field rendering on top of `@tanstack/react-form`, I am seeing the state of the form adding a new value on click and then immediately removing that item again only under `StrictMode`. I can reduce the problem enough that it occurs in once scenario but not another but its been hard to suss out exactly where the problematic code is coming from. 

I'll try to circle back around if I ever find that cause. Thanks for looking into this though!

**Comment by crutchcorn** [comment]
@dlindahl what version are you using? The most recent patches issues with this.
